pkg/ipamd/ipamd.go

```diff
@@ -499,17 +500,26 @@ func (c *IPAMContext) nodeInit() error {
 		}, 30*time.Second)
 	}
 
+	// Make a k8s client request for the current node so that max pods can be derived
+	node, err := k8sapi.GetNode(ctx, c.k8sClient)
```

Comment:
We should ideally remove the need of k8sClient usage unless any feature like custom networking would need it. We can enhance CNINode for this information. /cc @haouc

Reply:
Yeah, I thought about that, but we already have a non-optional k8s client call for the node when we attach an ENI - to check if the node is currently marked as schedulable. It would be nice to have this node state available in `CNINode`, but that would still require a k8s client call to retrieve.

Reply:
sure..with secondary IPs we would never go over max pods since max pods take secondary IP into account. This would be beneficial for prefix delegation but it is still tricky since allocation happens in chunks of 16..

Reply:
Right, that's we don't enforce a hard check on increase (or decrease). We simply do not increase, even when the warm targets tell us to, when max pods has been reached

Reply:
Also, in PD mode - Let's say we have WARM_PREFIX set to `1` and the node is already @ Max Pods. Now, with this change, we will not be honoring the WARM PREFIX in this state as the node is already at capacity w.r.t maxPods. If the user now churns the pods on this node (Let's say rollout restart of all deployments) - IPs assigned to recently deleted pods can't be reused for next 30s and so the new pods will not have any IPs and we would need to pull in a new prefix anyways? Current workflow, would've assigned an additional prefix and the new pods would've consumed IPs form the additional prefix and we would've released the old prefix after the cool down period. Did I understand the new flow right?

Reply:
Your understanding is correct. Over-provisioning could help in a case where the user restarts all pods at once. I do not think that is a case that we should design for, though. The downsides of over-provisioning (cost and resource utilization) seem much higher than slower recovery after restarting all pods, which should be an extremely rare event.